utils/src/offset/copy.rs

```diff
@@ -0,0 +1,264 @@
+
+        let mut buf = vec![0_u8; mem::size_of::<TestStruct>()];
+        let s = buf.as_mut_ptr() as *mut TestStruct;
+
+        unsafe {
+            (*s).a = -9223372036854771632;
+            (*s).b = 9223372036854771632;
+            (*s).c = -9223372036854771632;
+            (*s).d = 9223372036854771632;
+            (*s).e = -2147483623;
+            (*s).f = 2147483623;
+            (*s).g = -32721;
+            (*s).h = 32721;
+            (*s).i = -127;
+            (*s).j = 127;
+
+            let mut offset = offset_of!(TestStruct, a);
+            assert_eq!(offset, 0);
+            assert_eq!(
+                read_value_at::<isize>(&buf, &mut offset),
+                -9223372036854771632
+            );
+            assert_eq!(offset, 8);
+
+            let mut offset = offset_of!(TestStruct, b);
+            assert_eq!(offset, 8);
+            assert_eq!(
+                read_value_at::<usize>(&buf, &mut offset),
+                9223372036854771632
+            );
+            assert_eq!(offset, 16);
+
+            let mut offset = offset_of!(TestStruct, c);
+            assert_eq!(offset, 16);
+            assert_eq!(
+                read_value_at::<i64>(&buf, &mut offset),
+                -9223372036854771632
+            );
+            assert_eq!(offset, 24);
+
+            let mut offset = offset_of!(TestStruct, d);
+            assert_eq!(offset, 24);
+            assert_eq!(read_value_at::<u64>(&buf, &mut offset), 9223372036854771632);
+            assert_eq!(offset, 32);
+
+            let mut offset = offset_of!(TestStruct, e);
+            assert_eq!(offset, 32);
+            assert_eq!(read_value_at::<i32>(&buf, &mut offset), -2147483623);
+            assert_eq!(offset, 36);
+
+            let mut offset = offset_of!(TestStruct, f);
+            assert_eq!(offset, 36);
+            assert_eq!(read_value_at::<u32>(&buf, &mut offset), 2147483623);
+            assert_eq!(offset, 40);
+
+            let mut offset = offset_of!(TestStruct, g);
+            assert_eq!(offset, 40);
+            assert_eq!(read_value_at::<i16>(&buf, &mut offset), -32721);
+            assert_eq!(offset, 42);
+
+            let mut offset = offset_of!(TestStruct, h);
+            assert_eq!(offset, 44);
+            assert_eq!(read_value_at::<u16>(&buf, &mut offset), 32721);
+            assert_eq!(offset, 46);
+
+            let mut offset = offset_of!(TestStruct, i);
+            assert_eq!(offset, 48);
+            assert_eq!(read_value_at::<i8>(&buf, &mut offset), -127);
+            assert_eq!(offset, 49);
+
+            let mut offset = offset_of!(TestStruct, j);
+            assert_eq!(offset, 52);
+            assert_eq!(read_value_at::<u8>(&buf, &mut offset), 127);
+            assert_eq!(offset, 53);
+        }
+    }
+
+    #[test]
+    fn test_read_bounded_vec_at() {
+        #[derive(Clone, Copy, Pod, Zeroable)]
+        #[repr(C)]
+        struct TestStruct {
+            a: [i64; 32],
+            b: [u64; 32],
```

Comment:
probably makes sense to do an explicit test with [u8;32] as well since that is what we use in prod in most cases.

Reply:
Done